include/boost/multiprecision/cpp_double_fp.hpp

```diff
@@ -0,0 +1,2050 @@
+class cpp_double_fp_backend;
+
+template <typename FloatingPointType>
+constexpr void eval_add(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& x);
+template <typename FloatingPointType>
+constexpr void eval_add(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a, const cpp_double_fp_backend<FloatingPointType>& b);
+template <typename FloatingPointType>
+constexpr void eval_subtract(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& x);
+template <typename FloatingPointType>
+constexpr void eval_subtract(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a, const cpp_double_fp_backend<FloatingPointType>& b);
+template <typename FloatingPointType>
+constexpr void eval_multiply(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& x);
+template <typename FloatingPointType>
+constexpr void eval_multiply(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a, const cpp_double_fp_backend<FloatingPointType>& b);
+template <typename FloatingPointType>
+constexpr void eval_divide(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& x);
+template <typename FloatingPointType>
+constexpr void eval_divide(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a, const cpp_double_fp_backend<FloatingPointType>& b);
+template <typename FloatingPointType>
+constexpr bool eval_eq(const cpp_double_fp_backend<FloatingPointType>& a, const cpp_double_fp_backend<FloatingPointType>& b);
+template <typename FloatingPointType>
+constexpr bool eval_lt(const cpp_double_fp_backend<FloatingPointType>& a, const cpp_double_fp_backend<FloatingPointType>& b);
+template <typename FloatingPointType>
+constexpr bool eval_gt(const cpp_double_fp_backend<FloatingPointType>& a, const cpp_double_fp_backend<FloatingPointType>& b);
+template <typename FloatingPointType>
+constexpr bool eval_is_zero(const cpp_double_fp_backend<FloatingPointType>& x);
+template <typename FloatingPointType>
+constexpr int eval_signbit(const cpp_double_fp_backend<FloatingPointType>& x);
+
+template <typename FloatingPointType>
+constexpr void eval_fabs(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a);
+template <typename FloatingPointType>
+constexpr void eval_frexp(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a, int* v);
+template <typename FloatingPointType>
+constexpr void eval_ldexp(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a, int v);
+template <typename FloatingPointType>
+#if (defined(_MSC_VER) && (_MSC_VER <= 1900))
+BOOST_MP_CXX14_CONSTEXPR
+#else
+constexpr
+#endif
```

Comment:
Thanks Matt (@mborland) I'm doing a rather large rework of the numeric_limits for this work and I'll be doing a better job determining what is and is not really `constexpr` moving forward.

Reply:
There's one more detail we'll need to handle. MSVC 2015 (also known as VC 14.0) might have difficulties handling C++14: Relaxed constexpr restrictions. Subjectively, I find this to be the case when working with this compiler locally and in CI.

The _only_ compiler I'm using `BOOST_MP_CXX14_CONSTEXPR` for in my current PR drive is, in fact, VC 14.0. So if we drop that compiler, It'll be OK.

It's problem seems to be multiple return statements in a `constexpr` function. But if you really think of it, I'm not really super-100% sure if multiple return statements are allowed in a C++14 `constexpr` function?

So this might prove to be problematic when going C++14 as we move forward.